sys/fs/fuse/fuse_node.c

```diff
@@ -522,14 +528,17 @@ fuse_vnode_update(struct vnode *vp, int flags)
 	if (mp->mnt_flag & MNT_NOATIME)
 		flags &= ~FN_ATIMECHANGE;
 
+	if (flags) {
+		ASSERT_VOP_ELOCKED(vp, __func__); /* For fvdat->cached_attrs */
```

Comment:
Some other filesystems use the vnode interlock to serialize atime updates. Why can't that be done here? Disabling shared locking entirely seems like a heavy hammer.

Reply:
Probably, but I'm hazy on how the interlock works.  None of the man pages describe it.  How might I educate myself?

Reply:
For most purposes it's just a mutex(9) embedded in the vnode that can be used as an alternative to the vnode lock, especially for short code sections.

Within individual filesystems it's mostly up to the implementor to decide how to use it. There's a comment above the definition of struct vnode which explains how individual fields are locked; that's about all I'm aware of as far as documentation goes.

Reply:
More generically there is an interlock associated with lockmgr(9) locks, and the vnode interlock serves that purpose as well. Though, I can't find a general description of the "theory of operation" behind it.

Reply:
I'm trying that method, @markjdb .  However, I've also found another bug.  So I'm going to close this PR and split it up into 2-3 separate ones.